image/image_linux.go

```diff
@@ -451,14 +451,15 @@ func (s *imageSeeder) validateSnapArchs(snaps []*seedwriter.SeedSnap) error {
 
 type localSnapRefs map[*seedwriter.SeedSnap][]*asserts.Ref
 
-func (s *imageSeeder) deriveInfoForLocalSnaps(f seedwriter.SeedAssertionFetcher, db *asserts.Database) (localSnapRefs, error) {
+func (s *imageSeeder) deriveInfoForLocalSnaps(f seedwriter.SeedAssertionFetcher, db *asserts.Database, pathToLocalComp map[string]*snap.ComponentInfo) (localSnapRefs, error) {
 	localSnaps, err := s.w.LocalSnaps()
```

Comment:
should there be something like LocalComponents() as well?  we will symmetrically also be told what components to download to follow how the non-local case work?  There's a symmetry between LocalSnaps() and SnapsToDownload that we should try to follow

Reply:
I am trying to include information about the components in the snap structures, so `LocalSnaps()` should be enough as `SeedSnap` contains a slice of `SeedComponent`. The problem that I have here is that to fill that slice I have to do it in `deriveInfoForLocalSnaps` as in `w.SetOptionsSnaps` I cannot match local components to local snaps as I do not know the snap name until later, and I am trying to avoid a double read of the squashfs file.

This could be solved by some refactoring of the writer stages (running `w.Start` before `w.SetOptionsSnap`), but I did not want to put this in this PR as it was big enough.

The `[]*SeedSnap` returned by `SnapsToDownload` would also include needed components inside as required.

Reply:
my point is a bit different is not clear that we shouldn't let the writer manage something like pathToLocalComp internally? and let it do a bit more of the matching work as well, while letting the caller do a smaller part of the work. What I'm saying is that maybe SetOptionsSnaps should also take a list of OptionsComponent?

Reply:
Ok, I've pushed now more things to the writer and passed down pathToLocalComp so it can be used when we know the names of the local snap files. I think that passing as OptionComponent would complicate things a bit as then we would need to get back the local components to `image` to read the files, and also, we have already consumed the non-local related options.